### DISCUSSION TITLE
A rust version of the `KdPrint` macro

### DISCUSSION THREAD

**Discussion by mhandb** [microsoft/windows-drivers-rs#17] [Ideas]
In C, the `KdPrint` macro is convenient to add trace only in checked builds. 
 
In Rust, we could have something like this:
```
#[cfg(debug_assertions)]
#[macro_export]
macro_rules! kd_print {
    ($msg: expr) => {
        #[allow(unused_unsafe)]
        unsafe {
            DbgPrint(concat!($msg, "\0").as_ptr() as *mut i8)
        }
    };

    ($format: expr, $($arg:tt)*) => {
        #[allow(unused_unsafe)]
        unsafe {
             DbgPrint(concat!($format, "\0").as_ptr() as *mut i8, $($arg)*)
        }
    };
}

#[cfg(not(debug_assertions))]
#[macro_export]
macro_rules! kd_print {
    ($msg: expr) => {};
    ($format: expr, $($arg:tt)*) => {};
}
```
Which could be used this way
```
kd_print!("Log some value: %wZ\n", &some_unicode_string);
```

Here I kept the DbgPrint string formatting instead of relying on the Rust one. 
The purpose is to be able to easily log a `UNICODE_STRING`. 
But having a  UNICODE_STRING type which is displayable with `println!` would be better.

Having something similar which is built-in the crate, would be nice, at least for me :-) .

**Comment by wmmc88** [comment]
The `wdk` crate with the `alloc` feature enabled does actually [provide implementations](https://github.com/microsoft/windows-drivers-rs/blob/e127a84a90e63d4f124177960d45785d78839ded/crates/wdk/src/print.rs#L18) for `println` and `print` that use `alloc` features to print via `DbgPrint`. Albeit, using `alloc` features in their given state have some unresolved safety considerations per discussions in #5 and #6. 

> But having a UNICODE_STRING type which is displayable with println! would be better.

I agree. When we get to wrappers for `UNICODE_STRING`, something we could do is implement [`fmt::Display`](https://doc.rust-lang.org/core/fmt/trait.Display.html) for the type.
